부산대 BE 정윤성 - Step 2: 회원 로그인

dev-yunseong opened the pull request:
## 회원 로그인
  * 회원
    * DB Table 생성
    * Repository 생성
  * password hashing 해서 저장
    * password를 안전하게 관리하기 위해서 Bcrpt로 비가역적인 hashing을 해서 처리하도록 했습니다.
  * token 생성 
    * jjwt를 사용해서 토큰을 생성하도록 구현했습니다.
  * 회원가입
    * email의 중복을 먼저 확인 한 후에, Member를 저장하도록했습니다.
      * 이유
        * DB에서 UNIQUE 제한조건으로 인해 던지는 Exception이 h2만의 것이기에 확장성 측면에서 문제가 있어보였습니다.
        * 또한 실제 서비스들에서 회원가입에 앞서 id, email에 관해 중복 검사를 하기 때문에 이렇게 구현을 했습니다.
        * 또한 DB의 제한조건 검증은 최후의 보루로써 존재해야한다고 생각해서, 먼저 Service layer에서 중복 검증을 한 후에 저장하도록 했습니다.
  * 로그인
    * email, password 비교
    * email을 통해서 database에서 Member를 찾고 hashing 된 password와 평문 password를 Bcrpt를 통해 확인하도록 했습니다.
## 테스트 코드 구현
  * Product Controller를 위한 e2e 테스트
  * Member Controller를 위한 e2e 테스트

dev-yunseong committed:
Initialize Project - migration from "spring-gift-products repository"

dev-yunseong committed:
refacter ProductRepository -
* MemoryProductRepository.java - 사용되지 않는 class 삭제
* SQL Query String을 정적으로 선언

dev-yunseong committed:
doc README - step1의 기능 요구 사항 최신화

dev-yunseong committed:
feature improve Validation
- name의 최대 길이 수정
- 문자, 숫자, 특정 특수문자만 가능하도록 Pattern 설정
- BindingResult로 예외 메시지 생성해주는 RequestNotValidException 구현
- Controller에 Validation 예외 터르리는 부분 연결
- Exception Handler로 코드 중복을 줄이고 적절한 반응하도록 연결

dev-yunseong committed:
refactor schema Product.name length - validation에서 name을 15자로 제한함에 따라 schema에도 적용

dev-yunseong committed:
feature products-management.js - Product 생성, 수정때 - response에 따라서, 분기 구현
- onSuccess - modal창 끄기, reload
- onFailure - alert 띄우기

dev-yunseong committed:
feature Product에 Status 추가 -
* data.sql에 Status 추가
* schema.sql에 Status enum 추가, table에 attribute 추가
* ProductRepository - Status 추가, Status 수정 쿼리, 메서드 추가
* Product.Status enum 추가
* Product~Dto Status 추가
* ProductService - Status 자동 삽입 추가 ["카카오"가 들어있으면 Pending으로 기록], Approve 된 Products들만 가져오는 메서드 구현

dev-yunseong committed:
feature Product의 Status만 수정하는 Endpoint 구현 -
* DatabaseProductRepository - updateStatus에 예외처리, status Object가 아닌 String으로 넘기게 수정
* product의 status만 수정하는 endpoint 구현
  * 위를 위한 ProductStatusPatchRequestDto 구현
* ProductService status를 수정하는 메서드 추가
* README 수정 [admin 즉 상품을 등록하는 자와 승인하는 md의 책임이 다르다고 판단하여 수정]

dev-yunseong committed:
Merge branch 'dev-yunseong' into step1

dev-yunseong committed:
chore delete MemoryProductRepository.java - 필요 없는 class 삭제 [conflict 해결 중에 다시 되돌아옴]

dev-yunseong committed:
refactor Exception Handlers -
* global Exception Handler 수정
* 커스텀 예외가 아닌 MethodArgumentNotValidException를 사용해 처리하도록 수정

dev-yunseong committed:
doc step2 기능 요구 사항로 최신화

dev-yunseong committed:
feature add members table into db schema

dev-yunseong committed:
feature MemberRepository - save member and find member by email

dev-yunseong committed:
feature MemberService
- BCrpt를 통해 hashing 하여 Repository에 전달
- BCrpt을 통해 인증 구현
- Jwt 토큰을 생성, 반환하도록 구현
- Jwt의 Secret Key는 application.yml에 숨기도록 구현

dev-yunseong committed:
feature MemberController
- 회원가입 로그인에 대한 end point 구현
- 인증 관련 예외처리 handler 구현 [Login 할 때만 터질 것으로 기대 돼서, Member Controller에만 달아놓았습니다.]

dev-yunseong committed:
feature GlobalExceptionHandler - BadRequest에 대한 Exception handler 추가

dev-yunseong committed:
feature MemberControllerTest MemberController를 통한 end to end test 구현
- 회원 가입, 로그인 테스트 [성공적으로 회원가입, 로그인, password가 잘못 된 로그인을 처리하는지 테스트를 진행]
- 회원가입 중복 이메일 테스트 [성공적으로 중복 이메일 회원가입을 처리하는지 테스트]

dev-yunseong committed:
feature ProductControllerTest ProductController를 통한 end to end test 구현
- productRequestDto에 대한 validation 테스트
- 카카오가 들어가는 이름에 대한 처리 테스트
- 존재하지 않는 상품 읽기 테스트
- 상품 읽기 테스트

dev-yunseong committed:
Merge branch 'dev-yunseong' into step2

dev-yunseong committed:
feature conflict 해결중 다시 rollback된 부분 삭제

Gyuchool commented on README.md:
```diff
@@ -1,14 +1,15 @@
-  * Product에 Status를 수정할 수 있는 Endpoint 구현
-  * GET api/products 에선 APPROVED인 Product만 리턴하도록 수정
\ No newline at end of file
+* 회원
```
요구 사항 정리👍

Gyuchool commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,51 @@
+@RestController
+public class MemberController {
+
+    private final MemberService memberService;
+    private static final Logger log = LoggerFactory.getLogger(MemberController.class);
```
```suggestion
    private static final Logger log = LoggerFactory.getLogger(MemberController.class);
    
    private final MemberService memberService;

```

자바 관습적으로 상수는 맨 위에 적기는 합니다😃

dev-yunseong replied:
넵 반영하겠습니다!

Gyuchool commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,51 @@
+    private final MemberService memberService;
+    private static final Logger log = LoggerFactory.getLogger(MemberController.class);
+
+    MemberController(MemberService memberService) {
```
접근제어자 종류는 무엇들이 있나요?

dev-yunseong replied:
private, public 정도 있는 것 같습니다! 

private이라는 접근 제한자가 부적절한 것일까요??

Gyuchool replied:
22번 라인에 대해서 리뷰를 남기려했는데 19번 라인부터 남겨졌네요🙏

접근 제어자 종류와 어떻게 사용하는지, 역할이 무엇인지 설명해주실 수 있으실까요?

dev-yunseong replied:
보통 private, public, protected를 사용하는 것 같습니다.
또한 구글링을 통해 알아보니깐 아무것도 안 달았을 때는 java에선 private인줄 알았는데 default라는 패키지 단위의 접근 제한자인 것을 알게 되었습니다.

private class 안에서만 접근을 할 수 있고, 그렇기에 class 안에서만 사용되어야하는 필드에 달아주고 있습니다.
public은 모든 곳에서 접근을 할 수 있고, 그렇기에 객체 밖에서 호출해야하거나, 참조해야하는 곳에 달아주고 있습니다.
protected는 클래스를 상속 받는 자식 클래스에서도 접근할 수 있습니다. 그렇기에 추상화를 할 때, 자식 클래스들이 공통적으로 접근해야할 때 사용하고 있습니다.

Spring에서 의존성 주입을 위한 생성자를 만들 때는 그러한 접근 제한자에 대해 생각을 못 해봤네요.
제 생각에는 개발자가 객체화 하는 것을 최대한 막고, spring framework가 reflection을 통해서 접근할 수 있는 최소한의 접근 제한자를 사용하면 좋을 것 같습니다. 그게 default 일수도 있을 거 같습니다!

어떤 것을 사용하는게 좋을까요??

Gyuchool replied:
정해진 정답은 없습니다🤗 말씀대로 디폴트 접근제어자를 사용하셨길래 의도가 궁금했습니다:) 말씀주신대로 디폴트 접근제어자는 패키지 하위에서만 접근이 가능해서 만약 빈을 다른 패키지에서 생성하려면 실패하긴 하겠네요🤔
유연성을 위해서 public으로 생성해도 좋은데, 캡슐화/은닉의 관점에서는 default로 설정할 수도 있곘네요..:)

Gyuchool commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -38,12 +31,7 @@ public ResponseEntity<Void> deleteProduct(@PathVariable Long productId) {
     // For Admin
```
어떤 장점이 있어서 주석을 적용하셨을까요?

dev-yunseong replied:
이후 회원들의 Role에 따라 endpoint를 제한 할 때 구분하기 좋도록 적어준 주석입니다!

Gyuchool replied:
개인적으론 미래를 위한 주석은 좋지 않는것 같습니다..! 윤성님은 어떻게 생각하시나요?
주석의 장단점을 고민해보시고 의견 부탁드릴게요!

dev-yunseong replied:
주석은 코드를 설명해주지만 개발자가 직접 수정을 해야하니깐 실제 코드와의 동기화 측면에서 문제가 있을 것 같습니다.
또한 미래를 위한 코드는 협업자가 볼 때 어떤 걸 말하는 건지 헷갈릴 수도 있겠네요

미래를 위한 주석은 TODO 키워드를 통해서 명확히 해야할 일을 적어두는게 나을까요?? 아니면 아예 안 쓰는게 나을까요??

Gyuchool replied:
TODO가 있지만 구현되지않고 남아있는 여럿 코드들을 보아서(TODO가 언제 구현될지 모르거나 그 사이에 요구사항이 바뀌어서 계속 TODO로 남게되는 경우)...개인적으론 안좋아하긴 합니다😅
협업하다보면 팀원간에 주석에 대해서 의견이 달라질 수 있을텐데 그때 윤셩님의 기준이 있으면 좋을 것 같아서 질문드렸습니다~!

Gyuchool commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,51 @@
+        );
+    }
+
+    @ExceptionHandler(AuthenticationException.class)
```
https://github.com/next-step/spring-gift-wishlist/pull/88#discussion_r2180017410
이전 리뷰와 같은 의견입니당😃

dev-yunseong replied:
넵 반영하겠습니다!

보통 Exception Handler는 Global로 모아두는게 관습일까요??

Gyuchool replied:
관습보단 역할에 맞게 나뉘었다고 보시면 될 것 같아요!
지금처럼 컨트롤러에 핸들러를 추가하는 것과 GlobalExceptionHandler를 사용하는 것은 각각 어떤 장단점이 있을까요?

dev-yunseong replied:
컨트롤러에 핸들러를 추가하면 exception handler를 mapping하는 리소스를 필요한 곳에만 할 수 있을 것 같습니다. 하나면 몇 바이트 안 되겠지만, 많아진다면 커질 수도 있을 것 같습니다!

GlobalExceptionHandler를 사용한다면 개발자가 예측하지 못한 곳에서 발생한 예외도 처리될 가능성이 높아질 것 같습니다. 또한 코드 중복이 줄어들 것 같습니다.

Gyuchool replied:
좋습니다! 결국 같은 Exception을 핸들링하는데 10개 20개의 컨트롤러에서 중복으로 핸들링하는 것 보단 한 곳에서 핸들링하는게 유지보수하기에도 좋겠네요👍

Gyuchool commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,46 @@
+    @ExceptionHandler(NotFoundByIdException.class)
+    public ResponseEntity<String> handleNotFoundByIdException(NotFoundByIdException e) {
+        log.trace(e.getMessage());
+        return ResponseEntity.status(HttpStatus.NO_CONTENT)
```
HttpStatus에서 2XX, 3XX, 4XX, 5XX 각각 어떤 상황에 사용할까요?🤔

dev-yunseong replied:
200 - 성공적일 때 
300 - redirect 할 때
400 - client 측 요청에 에러에서 있을 때 
500 - server 측에서 예외가 일어났을 때

찾는 것이 없을 때 No content 보다 bad request가 어울릴까요??

Gyuchool replied:
넵:) 아무래도 성공이 아닌 실패를 응답하는 것이니, bad Request가 더 좋아보이긴 하네요..!

dev-yunseong replied:
넵 반영해보겠습니다!

Gyuchool commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,71 @@
+        this.secretKey = secretKey;
+    }
+
+    public AuthResponseDto creatMember(MemberRequestDto memberRequestDto) throws BadRequestException {
```
BadRequestException을 상위로 올리는 이유가 있을까요? 어떤 장점으로 올리신걸까요?

dev-yunseong replied:
BadRequestException이 RuntimeException을 상속 받지 않기 때문에 어쩔 수 없이 이렇게 구현을 했습니다.
실시간 강의를 들을 때 따로 custom exception을 안 만드는게 좋다고 하셔서 원래 존재하는 exception에 맞춰서 구현을 했는데 이런 경우엔 커스텀으로 만들고 exception signiture를 빼는게 나을까요?

Gyuchool replied:
지금보니 Exception이 BadReequestException이군요..🙄

import org.apache.coyote.BadRequestException;
import org.apache.tomcat.websocket.AuthenticationException;

두 의존성을 사용해서 예외처리를 하신 이유가 있으실까요? 
IllegalArgumentException() 대신 BadRequestException을 사용한 이유가 궁금하네요..!

dev-yunseong replied:
뭔가 상황에 맞는 exception을 찾기 위해서 했던 것 같습니다!
별 이유는 없었습니다..
IllegalArgumentException을 사용해봐야겠네요
AuthenticationException은 도 IllegalArgumentException로 대체하는게 좋을까요??

Gyuchool commented on src/main/java/gift/repository/MemberRepository.java:
```diff
@@ -0,0 +1,48 @@
+                .param("email", email)
+                .param("password", password)
+                .update(keyHolder);
+        return Optional.ofNullable(keyHolder.getKey());
```
Optional.ofNullable()로 감싼 이유가 무엇일까요?

dev-yunseong replied:
key가 null 일 수 있다는 Ide의 경고가 있어서 그런 상황엔 Service에서 예외처리를 하기 위해서 Optional로 감쌌습니다!

이전에 email unique 체크를 했으면, 논리적으론 key가 null일 가능성이 없으니깐
이러한 null 안정성을 생각 안 해도 될까요??

Gyuchool replied:
아뇨! null 체크는 항상 중요하다고 생각해요 좋은 습관인 것 같습니다👍
다만 Optional은 언제 사용하는게 좋을지, Optional이 무엇일지 고민해보는 것도 좋을 것 같습니다:)

참고 자료 남겨둡니다😄
https://tecoble.techcourse.co.kr/post/2021-06-20-optional-vs-null/

dev-yunseong replied:
감사합니다~

Gyuchool commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Transactional(readOnly = true)
+    public AuthResponseDto authenticateMember(MemberRequestDto memberRequestDto) throws AuthenticationException {
```
인증은 무엇인가요? 
`authenticateMember()` 네이밍은 너무 포괄적으로 느껴지는데 어떠신가요?

dev-yunseong replied:
그렇네요! 수정해보겠습니다!

Gyuchool commented on src/main/java/gift/util/BindingResultUtil.java:
```diff
@@ -1,16 +1,9 @@
-
-    @Override
-    public String getMessage() {
+public abstract class BindingResultUtil {
```
추상클래스로 만드신 이유가 있을까요?🙄 
유틸 클래스로 네이밍을 지은 이유도 궁금합니다!

dev-yunseong replied:
static 메서드만 가지고 있기에 객체화를 못하게 하기 위해서 추상 클래스로 구현했습니다.

---
사실 util이라는 클래스 명이 모든 기능의 클래스 명을 대변해주는 마법의 말이라 생각해서, 지양하려고 하고 있습니다
이 클래스 명을 지을 때 고민했던 이름이
BindingResultHelper
BindingResultExceptionMessageMaker
ExceptionMessageMaker
ValidationExceptionHelper
등이 있었는데 가장 어울리는 클래스 명을 지었습니다.

---
혹시 추천해주실만한 클래스 명이 있을까요??
또 클래스 명이 길어도 기능을 자세히 설명해주는 class가 좋을까요? 아님 좀 짧은 클래스 명을 짓기 위해 노력해야할까요?

Gyuchool replied:
1. 추상 클래스로 만들면 다른 개발자들은 상속을 먼저 생각하게 되어서 상속한 코드가 있을지 찾을 것 같은데 어떠신가요? 
    - 추상 클래스를 사용하지 않고 객체를 생성하지 못하게 하는 것은 어떠신가요?🙄

2. 네이밍은 저도 항상 고민이되어서..ㅎㅎ util을 붙여도 명확해서 좋은것 같긴합니다:) 다만, util클래스로 지정하신 이유가 궁금했습니다! util 패키지안의 클래스는 어떤 클래스들이 들어가게 될까요? 
    - 클래스 명의 길이에 대해서는 `클린 코드`와 `프로그래머스이 뇌` 라는 책을 읽으시면 본인만의 기준을 만드는데 좋을 것 같긴합니다! 정답이 없는 문제라서요😁

Gyuchool commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,71 @@
+                .longValue();
+
+        return new AuthResponseDto(
+                makeJwtToken(new Member(
```
객체 생성은 비싼 연산에 속하는데 객체를 생성한 이유가 있으실까요? id, email만 넘겨도 충분해 보여서요!

dev-yunseong replied:
이후 Jwt에 이름, 나이 등 다른 정보를 담을 수 있을 거 같은데, 정보를 추가할 때마다 인자를 추가하는 것 보다, 객체에 필드를 추가하는게 나을 것 같다고 생각해서 이렇게 구현했습니다!

하지만 jwt token을 생성할 때마다 따로 기능도 없는 개체를 생성해서 인자로 넘겨주는 것은 자원 부분에서 손해일 수도 있을 것 같네요 

각 필드를 인자로 풀어서 받는게 나을까요??

Gyuchool replied:
좋은 의도로 구현하신 것 같습니다!👍 말씀대로 객체가 커지고 불필요한 인자가 추가된다면 구조를 바꿀 필요가 있어보이긴하네요:)

Gyuchool commented on src/test/java/gift/controller/ProductControllerTest.java:
```diff
@@ -0,0 +1,110 @@
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class ProductControllerTest {
```
테스트 코드가 깨지는데 수정 부탁드립니다!🙏 테스트 코드 작성 시도💯💯💯

<img width="320" alt="스크린샷 2025-07-05 오후 8 36 54" src="https://github.com/user-attachments/assets/493f728c-d2ad-4be3-8580-446a212c449d" />

dev-yunseong replied:
넵 수정해보겠습니다! 감사합니다!

Gyuchool reviewed (changes requested):
안녕하세요 윤성님!
주말동안 미션 진행하시느라 고생하셨습니다!! 💪
요구사항대로 잘 구현해주셨지만 고민하고 학습할만한 포인트가 있어서 RC드립니다👍

몇 가지 코멘트를 남겨두었으니 반영 및 의견 주시고 다시 요청주세요!🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)
끝까지 화이팅입니다💪

dev-yunseong commented:
주말인데 리뷰 감사드립니다~

dev-yunseong committed:
refactor schema - product_status를 enum을 사용하지 않고 String으로 저장하도록 구현 [코드, dbms 동기화를 위해서]

dev-yunseong committed:
refactor - Product.Status를 추론하는 로직을 Domain layer로 이동

dev-yunseong committed:
refactor codes 리뷰 반영
 - authenticateMember 메서드 명 login으로 수정
 - 필요없는 주석 삭제
 - ExceptionHandler Global하게 처리하도록 수정

dev-yunseong committed:
chore conflict 해결중 되돌아온 파일 삭제

dev-yunseong committed:
fix Product Test Code - 카카오가 들어가는 Product 처리 테스트 코드 올바르게 처리하도록 수정

dev-yunseong committed:
refactor bean 생성을 위한 생성자의 코드 스타일 통일

dev-yunseong commented:
안녕하십니까?
리뷰 반영을 했봤습니다!
1. schema - enum이 아닌 String으로 저장핟도록 수정했습니다.
2. Status를 추론하는 로직을 Domain Layer로 옮겼습니다.
3. authenticateMember 메서드 명을 login으로 수정했습니다.
4. 필요 없는 주석 삭제 했습니다.
5. ExceptionHandler를 Global하게 처리하도록 수정했습니다.
6. 테스트 코드 로직이 잘 못 된 부분을 수정했습니다.
7. spring framework bean들을 위한 생성자 코드 스타일을 통일 했습니다. - Autowired 생략, 접근 제한자 public으로 통일

Gyuchool commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,9 +1,29 @@
+        );
+    }
+
+    public static Status inferStatus(ProductRequestDto productRequestDto) {
```
dto가 도메인까지 안들어오게 해주세요! 왜 그러는게 좋을까요?

Gyuchool replied:
추가로 static 클래스를 제거할 수 없을까요? static클래스가 있으면 어떤 단점이 있을까요?
   - Product 클래스를 record로 선언하신 이유도 궁금합니다:) 일반 class와 record class 중 record class로 도메인을 구현하신 이유를 공유해주세용😃

dev-yunseong replied:
dto를 도메인으로 변환하는게 service의 책임이고
dto와 domain의 의존성이 커져서 그럴 것 같습니다.
하지만 service에서 dto를 다 풀어해쳤다가 domain으로 수정하는 것은 별로인 것 같고,
dto에서 .toDomain() 메서드를 만드는 것도 비슷한 책임 문제가 있을 것 같은데
어떻게 해결해야할까요?

---

별 의도는 없었습니다!
간단하게 class를 구현할 수 있다는 점에서 선택했던 것 같습니다!

---

static 메서드가 있으면 메모리와 캡슐화 측면에서 문제가 있을 것 같습니다.
static 메서드를 제거하면 아래 부분에서 Product.inferStatus를 사용할 수가 없게 되는데 그럼
이때는 id가 없긴 하지만 id를 null로 두고 domain으로 변환한 뒤에 repository로 넘겨주는게 좋을까요??

```
@Transactional
public class ProductService {

    private final ProductRepository productRepository;

    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    public Long saveProduct(ProductRequestDto productRequestDto) {
        return productRepository.saveProduct(
            productRequestDto.name(),
            productRequestDto.price(),
            productRequestDto.imageUrl(),
            Product.inferStatus(productRequestDto.name())
        );
    }
```

Gyuchool replied:
1. 도메인은 가장 의존도가 적은 상태를 유지해야 합니다. 그렇기에 요구사항이 자주 변경되는 외부 dto를 도메인이 의존하지 않도록 해야하는 것이죠!
2. 현재는 Product 객체를 사용하지 않고있어서 이렇게 나온 것 같아요! serviceLayer에서 Product객체를 만들면 해소 될 것 같은데 한번 적용해보시고 궁금한 점이 추가로 나온다면 다음 단계에서 질문 주세욤!

dev-yunseong replied:
넵 감사합니다!

Gyuchool commented on src/main/resources/static/sql/schema.sql:
```diff
@@ -1,9 +1,13 @@
+CREATE TABLE members (
+    id BIGINT PRIMARY KEY AUTO_INCREMENT,
+    email VARCHAR(30) NOT NULL UNIQUE,
+    password CHAR(60) NOT NULL
```
password는 CHAR로 지정하신 이유가 있으실까요? CHAR, VARCHAR 차이가 무엇일까요?

dev-yunseong replied:
Char는 고정된 길이의 문자열이라 뒤쪽을 다 빈칸으로 채우는 것으로 알고 있습니다!
Varchar 은 최대 길이만 지정하고 저장된 길이의 문자열만들 저장한다고 알고 있습니다!
또한 고정된 길이로 인해 속도가 좀 더 빠르다고 알고 있습니다!

여기서 Bcrypt를 통해 얻은 hash를 저장하는데 고정된 60자로 나온하고 하여서 다음과 같이 저장했습니다

Gyuchool reviewed (approved):
2단계 미션 진행하시느라 고생하셨습니다!! 👍
피드백 반영도 잘해주시고 윤성님의 의견도 들려주셔서 더 좋았네요!💯💯

몇 가지 코멘트를 남겨두었으니 다음 미션 진행하시기전에 반영 및 의견 주시면 좋을 것 같습니다.🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!(새벽도 가능!)
다음 미션도 화이팅입니다!! 💪

Gyuchool commented:
> 안녕하십니까? 리뷰 반영을 했봤습니다!
> 
> 1. schema - enum이 아닌 String으로 저장핟도록 수정했습니다.
> 2. Status를 추론하는 로직을 Domain Layer로 옮겼습니다.
> 3. authenticateMember 메서드 명을 login으로 수정했습니다.
> 4. 필요 없는 주석 삭제 했습니다.
> 5. ExceptionHandler를 Global하게 처리하도록 수정했습니다.
> 6. 테스트 코드 로직이 잘 못 된 부분을 수정했습니다.
> 7. spring framework bean들을 위한 생성자 코드 스타일을 통일 했습니다. - Autowired 생략, 접근 제한자 public으로 통일

많은 것을 개선해보셨네요! 이를통해 많은 것을 얻어가셨으면 좋겠습니다!! 다음단계에서 뵐게요!
